dm/pkg/binlog/pos_finder.go

```diff
@@ -0,0 +1,369 @@
+			// TODO maybe add backtrace to support this case later
+			ev2, err := r.parser.Parse(ev.RawData)
+			if err != nil {
+				return false, err
+			}
+			e := ev2.Event.(*replication.QueryEvent)
+			switch string(e.Query) {
+			case "BEGIN":
+				transactionBeginEvent = true
+				r.inTransaction = true
+			case "COMMIT": // MyISAM use COMMIT to end transaction
+				r.inTransaction = false
+			default:
+				if !r.inTransaction {
+					// DDL
+					transactionBeginEvent = true
+				}
+			}
+		}
+	case replication.XID_EVENT:
+		r.inTransaction = false
+	}
+	return transactionBeginEvent, nil
+}
+
+// FindByTs get binlog location of first event or transaction with timestamp >= ts
+// go-mysql has BinlogStreamer.GetEventWithStartTime, but it doesn't fit our need. And we need to support relay log.
+// if posType != AboveUpperBoundBinlogPos, then location is the target location we want.
+// if posType == BelowLowerBoundBinlogPos, master binlog may have purged.
+func (r *binlogPosFinder) FindByTs(ts int64) (*Location, PosType, error) {
+	r.tctx.L().Info("target timestamp", zap.Int64("ts", ts))
+
+	if err := r.initTargetBinlogFile(ts); err != nil {
+		return nil, InvalidBinlogPos, err
+	}
+
+	targetTs := uint32(ts)
+	position := mysql.Position{Name: r.targetBinlog.name, Pos: FileHeaderLen}
+	gtidSet := gtid.MinGTIDSet(r.flavor)
+
+	binlogReader, err := r.startSync(position)
+	if err != nil {
+		return nil, InvalidBinlogPos, err
+	}
+	defer binlogReader.Close()
+	for {
+		ev, err := binlogReader.GetEvent(r.tctx.Ctx)
+		// let outer layer retry
+		if err != nil {
+			return nil, InvalidBinlogPos, err
+		}
+		if utils.IsFakeRotateEvent(ev.Header) {
+			continue
+		}
+
+		transactionBeginEvent, err := r.checkTransactionBeginEvent(ev)
+		if err != nil {
+			return nil, InvalidBinlogPos, err
+		}
+
+		if transactionBeginEvent && ev.Header.Timestamp >= targetTs {
+			break
+		}
+		position.Pos = ev.Header.LogPos
+
+		if r.enableGTID {
+			eventType := ev.Header.EventType
+			if eventType == replication.PREVIOUS_GTIDS_EVENT ||
+				eventType == replication.MARIADB_GTID_LIST_EVENT ||
+				eventType == replication.GTID_EVENT ||
+				eventType == replication.MARIADB_GTID_EVENT {
+				gtidSet, err = r.processGTIDRelatedEvent(ev, gtidSet)
+				if err != nil {
+					return nil, InvalidBinlogPos, err
+				}
+				// we meet PREVIOUS_GTIDS_EVENT or MARIADB_GTID_LIST_EVENT first, so break after get previous GTIDs
+				if r.tsBeforeFirstBinlog {
+					break
+				}
+			}
+		}
+
+		// still not found the timestamp after reached the end of this binlog file
+		if int64(position.Pos) >= r.targetBinlog.size {
+			// if it's the last binlog file, then this timestamp is out of range,
+			// else the end of this binlog file is the position we want,
+			// since the timestamp of the first event in next binlog >= target timestamp
+			if r.lastBinlogFile {
+				return nil, AboveUpperBoundBinlogPos, nil
+			}
+			break
+		}
+	}
+	if r.tsBeforeFirstBinlog {
```

Comment:
can we handle `r. tsBeforeFirstBinlog is true`  before above for loop ?

Reply:
we need to parse `PREVIOUS_GTIDS_EVENT` to get `gtidSet`